### PR TITLE
fix: fill schema-added nested columns with typed NULL arrays on read

### LINKED ISSUE #2618: Data files persisted without complex columns in the table schema hit "unexpected target column" on read
### Apache Iceberg Rust version

None

### Describe the bug

If we write a data file and then evolve our table's schema to add a complex type field like list, map, struct, proceeding to read the data file back will error out.

### To Reproduce

1. Create a table and write a data file (without column xs).
2. Evolve the schema to add xs list<int> (or a map/struct column).
3. Read the table.
4. Reading the older file errors with unexpected target column type List(...).

### Expected behavior

A column present in the table schema but absent from a data file is filled with a typed NULL array of the correct (possibly nested) Arrow type via new_null_array, for nested types just as for primitives. The read should succeed with the added column all-NULL for older files.

### Willingness to contribute

None

### PR DESCRIPTION
**Pull request by viirya**
## Which issue does this PR close?

- Closes #2618.

## What changes are included in this PR?

When a column of a nested type — `list`, `map`, or a struct that itself contains nested children — is added to the table schema after data files were written, reading those older files fails with `unexpected target column type List(...)`. The transformer correctly plans a `ColumnSource::Add { value: None, .. }` for the missing column, but the helpers that materialize the all-NULL array (`create_primitive_array_repeated` and `create_primitive_array_single_element` in `arrow/value.rs`) only covered primitive types plus structs with primitive-only children, each via a hand-written per-type NULL branch.

This PR replaces all of those NULL branches with a single early return using arrow's `new_null_array`, which constructs a typed all-NULL array for **every** Arrow type, including arbitrarily nested ones (the timezone of timestamps and precision/scale of decimals are part of the `DataType`, so they are preserved). The `Some(literal)` branches — used for `initial_default` values and partition constants — are unchanged. Net effect: the two functions shrink by ~180 lines and the unsupported-type failure mode for NULL filling disappears entirely.

## Are these changes tested?

New regression test `schema_evolution_adds_list_map_and_nested_struct_columns_with_nulls` in `record_batch_transformer.rs`: a file batch containing only `id` is read against an evolved schema that added `xs: list<int>`, `props: map<string, int>`, and `s: struct<a: string, ys: list<long>>` (the struct's `ys` child also exercises the nested-children path that the old `Struct` branch couldn't handle). The test asserts the added columns come back with the evolved schema's Arrow types and `null_count == num_rows`.

The test fails on main with `unexpected target column type List(Int32, ...)` — the exact error from the issue — and passes with this change. Full `iceberg` lib suite (1313 tests) passes; clippy and rustfmt clean.

### CODE REVIEW

File: crates/iceberg/src/arrow/value.rs [hunk 1]
```diff
@@ -628,23 +627,25 @@ pub(crate) fn create_primitive_array_single_element(
     data_type: &DataType,
     prim_lit: &Option<PrimitiveLiteral>,
 ) -> Result<ArrayRef> {
+    // With no value, the single element is NULL. `new_null_array` supports every
+    // Arrow type, including nested ones (list/map/struct), which matters for
+    // columns added by schema evolution after a data file was written (#2618).
+    if prim_lit.is_none() {
```
Comment (advancedxy) [inline crates/iceberg/src/arrow/value.rs:632]: Hi @viirya thanks for the fix. We encountered the same issue with a schema evolution case: a list of binary was added, and the original code cannot handle it. We fixed that internally and are about to contribute it back. 

Our fix is similar with yours. However during the internal code review process, we noticed the function name `create_primitive_array_single_element` is no longer valid, it's not just creating primitive array any more, it  also creating list/struct/map arrays now. Considering the default value in Iceberg V3, a struct could have default values. I think it would be best to change the signature to create_array_single_element snd passes prim_lit as `&Option<Literal>`, WDYT?

Comment (advancedxy) [reply crates/iceberg/src/arrow/value.rs:632]: I think this fix is valid short-term fix. We can get it merged first, and refactor it in a follow-up PR to address the naming issue and default value support in Iceberg V3.

Comment (viirya) [reply crates/iceberg/src/arrow/value.rs:632]: Thanks @advancedxy — agreed on both counts, and thanks for flagging the V3 angle.

You're right that `create_primitive_*` is now a misnomer since these functions materialize list/map/struct NULLs too. I dug into the V3 default-value direction a bit, and it's actually a slightly larger refactor than just the helper signatures: the value currently threaded into them is `Option<PrimitiveLiteral>` (via `ColumnSource::Add`), and `generate_transform_operations` deliberately drops any non-primitive `initial_default` today (`if let Literal::Primitive(prim) = lit { .. } else { None }` in `record_batch_transformer.rs`). So supporting a struct/nested default would mean widening `ColumnSource::Add.value` to `Option<Literal>` and the transformer's default-extraction alongside the rename to `create_array_single_element`.

Given that, I'd prefer to keep this PR as the focused wraparound/nested-NULL fix and do the rename + `Literal` widening + V3 default support together in the follow-up, so the type change lands in one coherent step rather than renaming now and re-touching the signature later. Happy to open the follow-up issue/PR for that.

Comment (advancedxy) [reply crates/iceberg/src/arrow/value.rs:632]: > Given that, I'd prefer to keep this PR as the focused wraparound/nested-NULL fix and do the rename + Literal widening + V3 default support together in the follow-up

Agreed. Let's merge this first.

**Review by advancedxy — APPROVED** [review]

**Comment by viirya** [comment]
cc @kevinjqliu

File: crates/iceberg/src/arrow/record_batch_transformer.rs [hunk 2]
```diff
@@ -921,6 +923,126 @@ mod test {
         assert!(struct_column.is_null(2));
     }
 
+    #[test]
+    fn schema_evolution_adds_list_map_and_nested_struct_columns_with_nulls() {
```
Comment (kevinjqliu) [inline crates/iceberg/src/arrow/record_batch_transformer.rs]: nit: this test is a bit difficult to read. do you mind ordering so its easier to decipher the test? 

we're using an evolved schema (with new optional list/map/struct types) to read the a data file that has the older schema. And we expect that all new optional list/map/struct types to be null. 

should we also test for the shape of the null value? e.g. the null struct still has the expected child fields?

File: crates/iceberg/src/arrow/value.rs [hunk 3]
```diff
@@ -628,23 +627,25 @@ pub(crate) fn create_primitive_array_single_element(
     data_type: &DataType,
     prim_lit: &Option<PrimitiveLiteral>,
 ) -> Result<ArrayRef> {
+    // With no value, the single element is NULL. `new_null_array` supports every
+    // Arrow type, including nested ones (list/map/struct), which matters for
+    // columns added by schema evolution after a data file was written (#2618).
```
Comment (kevinjqliu) [inline crates/iceberg/src/arrow/value.rs]: nit: feels like we're over explaining here, wdyt?

Comment (kevinjqliu) [reply crates/iceberg/src/arrow/value.rs]: maybe we should document the `1`, i think its for `new_null_array` 's `num_rows`

File: crates/iceberg/src/arrow/value.rs [hunk 4]
```diff
@@ -653,14 +654,6 @@ pub(crate) fn create_primitive_array_single_element(
                 Ok(Arc::new(array))
             }
         }
-        (DataType::Timestamp(TimeUnit::Microsecond, timezone), None) => {
```
Comment (kevinjqliu) [inline crates/iceberg/src/arrow/value.rs:656]: just double checking, do you know if `new_null_array` handles this and the nanosecond case correctly?

File: crates/iceberg/src/arrow/value.rs [hunk 5]
```diff
@@ -721,81 +700,6 @@ pub(crate) fn create_primitive_array_single_element(
...
-                    Error::new(
-                        ErrorKind::DataInvalid,
-                        format!(
-                            "Failed to create Decimal128Array with precision {precision} and scale {scale}: {e}"
-                        ),
-                    )
-                })?;
-            Ok(Arc::new(array))
-        }
-        (DataType::Struct(fields), None) => {
```
Comment (kevinjqliu) [inline crates/iceberg/src/arrow/value.rs:737]: nice, glad we can get rid of these with the helper method!

File: crates/iceberg/src/arrow/value.rs
Comment (kevinjqliu) [inline crates/iceberg/src/arrow/value.rs:713]: maybe a good opportunity for refactoring with `create_primitive_array_repeated`, i see a lot of repetition. but its outside this pr's scope 😄

**Review by kevinjqliu — COMMENTED** [review]
Thanks! A couple questions from me. Would be good to have others take a look too

BTW I also notice there's another PR addressing the same issue (#2649)

File: crates/iceberg/src/arrow/value.rs (hunk 4 above)
Comment (mbutrovich) [reply crates/iceberg/src/arrow/value.rs:656]: Yes. `new_null_array` builds from the `DataType`, and for timestamps the timezone is part of `DataType::Timestamp(unit, tz)` (and precision/scale are part of `Decimal128`), so the null array carries the exact same type the old per-type branches produced, including `Timestamp(Nanosecond, tz)`. The thing it does not preserve is RunEndEncoded wrapping, but the NULL-fill path uses the plain schema type (REE is only applied on the `Some`/constant path via `datum_to_arrow_type_with_ree`), so there is no regression. A test asserting the full nested `DataType` (which kevinjqliu also asked for) would lock this in.

**Review by mbutrovich — APPROVED** [review]
Building on kevinjqliu's note that #2649 addresses the same issue: this is the one to take. #2649 is an additive per-type fix (+80/-1); this PR's `new_null_array` approach is more general, removes the existing per-type branches rather than extending them, and its test covers a struct whose child is itself a list (`s: struct<a, ys: list<long>>`) that an enumerate-each-type fix would miss. Recommend landing this and closing #2649 as redundant (the bug still gets fixed and #2649's intent is fully covered here).

Thanks @viirya!

**Comment by viirya** [comment]
Thanks @kevinjqliu — addressed the feedback in e845ae08:

- **Test readability / ordering.** Extracted the evolved schema into `schema_with_added_nested_columns()` and reordered the test body so it reads top-to-bottom as the story you described: (1) the old data file with only `id`, (2) read it against the evolved schema, (3) expect the added list/map/struct columns as typed NULLs. The big nested-schema definition no longer sits in front of the actual logic.

- **Null shape assertion.** Added an assertion that the all-NULL struct still carries its full nested shape — child fields `a` and the nested list `ys` — rather than collapsing to an empty struct. This is exactly the property the `new_null_array` approach guarantees over an enumerate-each-type fix (and what @mbutrovich noted).

- **`new_null_array` for timestamp/decimal (your question on the `Some` timestamp arm).** Confirmed: `new_null_array` builds from the `DataType`, and timezone is part of `DataType::Timestamp(unit, tz)` and precision/scale part of `Decimal128`, so the NULL array carries the exact type the old per-type branches produced. I verified locally:
  ```
  new_null_array(Timestamp(Nanosecond, Some("+08:00")), 1) -> dt = Timestamp(Nanosecond, Some("+08:00")), null
  new_null_array(Decimal128(20, 4), 1)                     -> dt = Decimal128(20, 4), null
  ```
  (REE wrapping isn't preserved, but the NULL-fill path uses the plain schema type — REE is only applied on the constant/`Some` path — so no regression.)

- **Comments.** Trimmed the over-explanatory comments on both NULL-fill paths and noted the literal `1` is `new_null_array`'s row count.

- **Refactoring `create_primitive_array_repeated` (the repetition you spotted) and the `create_primitive_*` rename / `Literal` widening for V3 default values** — agreed these are worth doing, keeping them as the follow-up @advancedxy and I discussed above so this PR stays the focused #2618 fix.

**Comment by viirya** [comment]
Thanks @mbutrovich for reviewing.

**Comment by viirya** [comment]
@kevinjqliu Could you take another look? Thanks!

**Comment by viirya** [comment]
@kevinjqliu @blackmwk or @CTTY can you take a look? Thanks!

File: crates/iceberg/src/arrow/value.rs [hunk 6]
```diff
@@ -937,34 +794,6 @@ pub(crate) fn create_primitive_array_repeated(
...
-                .map(|field| create_primitive_array_repeated(field.data_type(), &None, num_rows))
-                .collect::<Result<Vec<_>>>()?;
-
-            Arc::new(StructArray::new(
-                fields.clone(),
-                null_arrays,
-                Some(NullBuffer::new_null(num_rows)),
-            ))
-        }
         (DataType::Null, _) => Arc::new(arrow_array::NullArray::new(num_rows)),
```
Comment (laskoviymishka) [inline crates/iceberg/src/arrow/value.rs]: Now that the early return catches every `prim_lit.is_none()` case up top, this arm is only reachable with `Some(_)` — a bare `Null` type with no literal already returns above. Harmless, but it reads as if it still handles the None case. I'd tighten it to `(DataType::Null, Some(_))` or just drop it. wdyt?

File: crates/iceberg/src/arrow/record_batch_transformer.rs [hunk 7]
```diff
@@ -921,6 +923,150 @@ mod test {
...
+                            false,
+                        )
+                        .into(),
+                    }),
+                )
+                .into(),
+                NestedField::optional(
+                    7,
+                    "s",
+                    Type::Struct(crate::spec::StructType::new(vec![
```
Comment (laskoviymishka) [inline crates/iceberg/src/arrow/record_batch_transformer.rs:1495]: `s` is added here with no `initial_default`, so null-fill is exactly right. The case this PR quietly makes reachable is a struct added with a non-primitive `initial_default` (e.g. `{}`) — the `and_then` in the transformer drops it to `None`, so we'd null-fill instead of materializing the spec-mandated non-null struct. Not blocking, but a small pin-test here (even `#[ignore]`'d) documenting that gap would pair well with the follow-up I noted in the summary.

File: crates/iceberg/src/arrow/record_batch_transformer.rs [hunk 8]
```diff
@@ -921,6 +923,150 @@ mod test {
...
+
+        // (3c) The all-NULL struct still carries its full nested shape (`a` plus the
+        // nested list `ys`), not a degenerate empty struct — this is what the
+        // type-preserving NULL fill guarantees over an enumerate-each-type fix.
+        let result_schema = result.schema();
+        let DataType::Struct(struct_fields) = result_schema.field(3).data_type() else {
+            panic!("field `s` should be a struct");
+        };
+        let child_names: Vec<&str> = struct_fields.iter().map(|f| f.name().as_str()).collect();
+        assert_eq!(child_names, vec!["a", "ys"]);
```
Comment (laskoviymishka) [inline crates/iceberg/src/arrow/record_batch_transformer.rs:1585]: Nice — pinning that the null struct keeps `a` and `ys` (and that `ys` stays a List) is exactly the shape a per-type fix would flatten. Could we also assert the child fields keep their `PARQUET_FIELD_ID`s (8/9 here, plus 3/5/6 on the top-level list/map)? That metadata is the load-bearing guarantee of this approach and what downstream name-matching relies on, so it'd be good to guard against a refactor that strips it. wdyt?

**Review by laskoviymishka — APPROVED** [review]
agreed with @mbutrovich, this is the one to take. Collapsing the per-type `None` branches into `new_null_array` is the right generalization, and the struct-with-a-nested-list test (`s: struct<a, ys: list<long>>`) pins exactly the case a per-type fix would miss. Happy to see #2649 close as redundant.

One thing worth flagging, and it's explicitly not a blocker: now that the null path works for nested types, a newly-added column whose `initial_default` is a non-primitive `Literal` (struct/list/map) gets silently dropped to `None` by the `and_then` in `record_batch_transformer.rs` (~L872) and null-filled — where the spec says a struct added with `initial_default = {}` should read back as a non-null struct with each child's own default applied. It's a pre-existing gap, just more visible now, and the fix is out of scope here. I'd file a follow-up (materialize complex defaults, or reject them in `NestedField::with_initial_default` to match Java's `castDefault`) and drop a `// TODO` at that `and_then` so it doesn't get lost.

Left a couple of small inline nits too: the retained `(DataType::Null, _)` arm reads as if it still handles None, and it'd be nice to pin the child field-ids in the new test. 

None of that blocks; this is good to merge, will left soaking for some time before merging.

**Comment by viirya** [comment]
Thanks @laskoviymishka for the approval and the detailed feedback! Updated in cc794c82, which also merges the latest upstream main and resolves the conflicts.

I tightened the `Null` arm to `Some(_)` and added assertions for the nested field IDs, including the element of `s.ys`. I also added a TODO at the default extraction to document the existing complex-default gap. The implementation and regression coverage for complex defaults can stay together in a follow-up, consistent with the earlier discussion about widening the helpers to accept `Literal`.

All 1,675 iceberg library tests pass locally, along with clippy and formatting checks.

File: crates/iceberg/src/arrow/value.rs (hunk 6 above)
Comment (viirya) [reply crates/iceberg/src/arrow/value.rs]: Good point — changed this to `(DataType::Null, Some(_))`. This makes the reachable case explicit while preserving the existing behavior for `Null` with a supplied literal; dropping the arm would make that case return an error.

File: crates/iceberg/src/arrow/record_batch_transformer.rs (hunk 7 above)
Comment (viirya) [reply crates/iceberg/src/arrow/record_batch_transformer.rs:1495]: Agreed — the extraction currently drops non-primitive defaults to `None`. Added a TODO at that `and_then` to keep the gap visible. I'd keep the regression test with the complex-default implementation in a follow-up, where we can cover the non-null struct and child-default semantics together.

File: crates/iceberg/src/arrow/record_batch_transformer.rs (hunk 8 above)
Comment (viirya) [reply crates/iceberg/src/arrow/record_batch_transformer.rs:1585]: Added assertions for `PARQUET_FIELD_ID` on `a`/`ys` (8/9), the top-level list element (3), and the map key/value (5/6). I also included the nested `ys` element (10), so the test covers metadata preservation through that extra nesting level.

**Comment by laskoviymishka** [comment]
@viirya i think this is good to merge, fix the conflicts and we are good.

File: crates/iceberg/src/arrow/record_batch_transformer.rs
Comment (dannycjones) [inline crates/iceberg/src/arrow/record_batch_transformer.rs]: I hope to follow-up on this soon so not a big deal, but in general I'd prefer we fail closed if we cannot support this.

I was planning to introduce this:

```rust
let default_value = match iceberg_field.initial_default.as_ref() {
    None => None,
    Some(Literal::Primitive(prim)) => Some(prim.clone()),
    Some(_) => {
        return Err(Error::new(
            ErrorKind::FeatureUnsupported,
            format!(
                "Cannot read field {} that is absent from the data file: \
                    applying a non-primitive initial-default for {} is not yet \
                    supported",
                iceberg_field.name, iceberg_field.field_type
            ),
        ));
    }
};
```

i.e. this PR will take an obscure failure reading data files without the complex column, into implementing `null` filling for complex columns without a default and properly reporting a clear error for those with a default.

**Review by dannycjones — COMMENTED** [review]
thanks for working on this, I'd like to see this merged soon.

I haven't took a look at the full PR, just had one note on the behavior change

**Comment by viirya** [comment]
Thanks! Merged the latest upstream main and resolved the conflicts in fbc8f72b.

I also addressed @dannycjones’s suggestion: missing fields with non-primitive initial defaults now return `FeatureUnsupported` with a descriptive error, instead of silently producing NULLs. Added regression coverage for both optional and required structs.

File: crates/iceberg/src/arrow/record_batch_transformer.rs
Comment (viirya) [reply crates/iceberg/src/arrow/record_batch_transformer.rs]: Agreed — silently replacing an unsupported default with NULL could return incorrect data. Applied your suggested match in fbc8f72b: absent optional fields without a default still get typed NULLs, primitive defaults retain their existing behavior, and non-primitive defaults return `FeatureUnsupported` with the field name and type.

Added regression coverage for both optional and required structs with an initial default. The TODO now tracks implementing complex defaults, including child defaults.

Comment (dannycjones) [reply crates/iceberg/src/arrow/record_batch_transformer.rs]: Thanks for addressing this one!
